lock-api-objects/src/main/java/com/palantir/lock/watch/LockEvent.java

```diff
@@ -52,6 +62,65 @@ public <T> T accept(Visitor<T> visitor) {
+            @JsonProperty("metadata") Optional<WireLockRequestMetadata> optWireMetadata) {
+        Optional<LockRequestMetadata> metadata = optWireMetadata.map(wireMetadata -> {
+            KeyListChecksum checksum = KeyListChecksum.of(
+                    ChecksumType.valueOf(wireMetadata.checksumTypeId()), wireMetadata.checksumValue());
+            IndexEncodingResult<LockDescriptor, ChangeMetadata> indexEncodingResult =
+                    IndexEncodingResult.<LockDescriptor, ChangeMetadata>builder()
+                            .keyList(lockDescriptors)
+                            .indexToValue(wireMetadata.indexToChangeMetadata())
+                            .keyListChecksum(checksum)
+                            .build();
+            Map<LockDescriptor, ChangeMetadata> lockDescriptorToChangeMetadata =
+                    IndexEncodingUtils.decode(indexEncodingResult, Function.identity());
+            return LockRequestMetadata.of(lockDescriptorToChangeMetadata);
+        });
+        return ImmutableLockEvent.builder()
+                .sequence(sequence)
+                .lockDescriptors(ImmutableSet.copyOf(lockDescriptors))
+                .lockToken(lockToken)
+                .metadata(metadata)
+                .build();
+    }
+
+    @Unsafe
+    @JsonValue
+    public Map<String, Object> toJson() {
+        Optional<Pair<List<LockDescriptor>, WireLockRequestMetadata>> optLocksAndWireMetadata = metadata()
+                .map(metadata -> {
+                    IndexEncodingResult<LockDescriptor, ChangeMetadata> indexEncodingResult = IndexEncodingUtils.encode(
+                            lockDescriptors(),
+                            metadata.lockDescriptorToChangeMetadata(),
+                            Function.identity(),
+                            DEFAULT_CHECKSUM_TYPE);
+                    KeyListChecksum keyListChecksum = indexEncodingResult.keyListChecksum();
+                    WireLockRequestMetadata wireMetadata = WireLockRequestMetadata.builder()
+                            .indexToChangeMetadata(indexEncodingResult.indexToValue())
+                            .checksumTypeId(keyListChecksum.type().getId())
+                            .checksumValue(keyListChecksum.value())
+                            .build();
+                    return Pair.create(indexEncodingResult.keyList(), wireMetadata);
+                });
+        // If the lock event has metadata, we need to use the order that was used to encode the metadata.
+        // Otherwise, we can use any order.
+        List<LockDescriptor> orderedLocks =
+                optLocksAndWireMetadata.map(Pair::getLhSide).orElseGet(() -> ImmutableList.copyOf(lockDescriptors()));
+        ImmutableMap.Builder<String, Object> mapBuilder = ImmutableMap.<String, Object>builder()
+                .put("sequence", sequence())
+                .put("lockDescriptors", orderedLocks)
+                .put("lockToken", lockToken());
+        optLocksAndWireMetadata
```

Comment:
I have also considered using a proper object here instead of a `Map<String, Object>`. Something like `WireLockEvent` (which basically mirrors the `LockEvent` fields) or even creating a custom `JsonObject` using some jackson built-ins (if that is possible). The downside of using something like `WireLockEvent` is that the serialization is spread out across multiple classes and changes become even more involved. Also, we'd have to ensure that field names match exactly.

Reply:
Opted for the object-based approach with a new `WireLockEvent` class nested inside `LockEvent`